colm/transpile/transpile.lm

```diff
@@ -0,0 +1,265 @@
+{
+	# Add measurement as a filter.
+	for Measurement: measurement in SS.from_clause {
+		M: str = $Measurement
+		if match Measurement [unquoted_identifier]
+			M = "\"[M]\""
+		send Output
+			"	|> filter( fn: (r) => ( r._measurement == [M] ) )
+	}
+}
+
+void add_where_filter( Output: output, WC: where_clause )
+{
+	# We are hacking the influxql query so that it will parse as a flux query.
+	for Op: AND in WC.expr {
+		match Op [And: `AND]
+		And.data = 'and'
+		Op = cons AND [And]
+	}
+
+	for Op: OR in WC.expr {
+		match Op [Or: `OR]
+		Or.data = 'or'
+		Op = cons OR [Or]
+	}
+
+	for CE: comparison_operator in WC.expr {
+		if match CE [Eq: `=] {
+			Eq.data = "=="
+			CE = cons comparison_operator [ Eq ]
+		}
+	}
+
+	for SL: string_lit in WC.expr {
+		Str: str = $SL
+		Str = prefix( suffix( Str, 1 ), Str.length - 2 )
+		SL.data = "\"[Str]\""
+	}
+
+	# Reparsing the expression as a flux query. This sends the InfluxQL tree,
+	# as text, to a flux expession parser.
+	parse Expr: flux::expression [WC.expr]
+	if !Expr {
+		send stderr
+			"ERROR: failed to reparse remaining parts of where clause as flux expression: [error]
+			"ERROR: [WC.expr]
+		exit(1)
+	}
+	
+	# After reparsing as flux we do some rewriting to add r. qualifications.
+	for UE: flux::unary_expression in Expr {
+		if match UE [ID: flux::identifier] {
+			UE = cons flux::unary_expression " r.[ID] "
+		}
+	}
+
+	# Generate.
+	send Output
+		"	|> filter( fn: (r) => ( [Expr ] ) )
+}
+
+void add_select_filter( Output: output, SS: select_stmt )
+{
+	# Add select expression as a filter. Note that this just works on simple
+	# select expression that name a value. It does not evaluate expressions.
+	for Field: field in SS.fields {
+		if match Field [Id: identifier] {
+			FieldName: str = $Id
+			if match Id [unquoted_identifier]
+				FieldName = "\"[FieldName]\""
+
+			send Output
+				"	|> filter( fn: (r) => ( r._field == [FieldName] ) )
+		}
+	}
+}
+
+# (not an exhaustive list)
+# 1. Extract time constraints from the where clause.
+# 2. Build range from extracted time constraints.
+# 3. Add measurement filter.
+# 4. Build filter from remaining where clause.
+# 5. Apply group by (not implemented)
+# 6. Apply simple select expressions.
+
+alias output parser<flux::flux>
+
+for SS: select_stmt in P {
+	# send stderr "select statement: [^SS]
+	cons Range: time_range []
+
+	# Extracting time range from the whre clause. Leaves WC.expr altered.
+	WC: where_clause = SS.opt_where_clause.where_clause
+	if WC {
+		yank_time( Range, WC.expr )
+		clean_expr( WC.expr )
+	}
+
+	# If there is no start range, add one. 
+	if !Range.Start
+		Range.Start = parse flux::expression "-1m"
+
+	# Writing output. Note that this output stream is a parser, so it is
+	# parsing what is sent to it, as we go along. At the end we will extract
+	# the tree and there is opportunity for further rewriting on that. Although
+	# we are not doing that in this version.
```

Comment:
This is really cool. So currently you're walking an influxQL tree and piping flux source through a parser to get the final flux syntax tree. You could also have output flux syntactic elements directly from influxQL elements. What's the benefit of the first method (sending flux source through a parser) vs the second? It seems like the second method gives a compile time guarantee that you're producing syntactically correct flux. Is this true?

Reply:
It's simply faster to implement. This part I did in under 30 minutes because I had burned up all my time box figuring out time constraints in InfluxQL.

Exactly correct. It's better to build up with proper types in a bottom up fashion (you can also go top down if you want). Using the hack above it's quick, but it's also risky.

Reply:
Thank you for the great questions!